libevmasm/SimplificationRule.h

```diff
@@ -55,84 +56,106 @@ struct SimplificationRule
-	static auto constexpr CALLDATASIZE = Instruction::CALLDATASIZE;
-	static auto constexpr CALLDATACOPY = Instruction::CALLDATACOPY;
-	static auto constexpr CODESIZE = Instruction::CODESIZE;
-	static auto constexpr CODECOPY = Instruction::CODECOPY;
-	static auto constexpr GASPRICE = Instruction::GASPRICE;
-	static auto constexpr EXTCODESIZE = Instruction::EXTCODESIZE;
-	static auto constexpr EXTCODECOPY = Instruction::EXTCODECOPY;
-	static auto constexpr RETURNDATASIZE = Instruction::RETURNDATASIZE;
-	static auto constexpr RETURNDATACOPY = Instruction::RETURNDATACOPY;
-	static auto constexpr EXTCODEHASH = Instruction::EXTCODEHASH;
-	static auto constexpr BLOCKHASH = Instruction::BLOCKHASH;
-	static auto constexpr COINBASE = Instruction::COINBASE;
-	static auto constexpr TIMESTAMP = Instruction::TIMESTAMP;
-	static auto constexpr NUMBER = Instruction::NUMBER;
-	static auto constexpr DIFFICULTY = Instruction::DIFFICULTY;
-	static auto constexpr GASLIMIT = Instruction::GASLIMIT;
-	static auto constexpr CHAINID = Instruction::CHAINID;
-	static auto constexpr SELFBALANCE = Instruction::SELFBALANCE;
-	static auto constexpr POP = Instruction::POP;
-	static auto constexpr MLOAD = Instruction::MLOAD;
-	static auto constexpr MSTORE = Instruction::MSTORE;
-	static auto constexpr MSTORE8 = Instruction::MSTORE8;
-	static auto constexpr SLOAD = Instruction::SLOAD;
-	static auto constexpr SSTORE = Instruction::SSTORE;
-	static auto constexpr PC = Instruction::PC;
-	static auto constexpr MSIZE = Instruction::MSIZE;
-	static auto constexpr GAS = Instruction::GAS;
-	static auto constexpr LOG0 = Instruction::LOG0;
-	static auto constexpr LOG1 = Instruction::LOG1;
-	static auto constexpr LOG2 = Instruction::LOG2;
-	static auto constexpr LOG3 = Instruction::LOG3;
-	static auto constexpr LOG4 = Instruction::LOG4;
-	static auto constexpr CREATE = Instruction::CREATE;
-	static auto constexpr CALL = Instruction::CALL;
-	static auto constexpr CALLCODE = Instruction::CALLCODE;
-	static auto constexpr STATICCALL = Instruction::STATICCALL;
-	static auto constexpr RETURN = Instruction::RETURN;
-	static auto constexpr DELEGATECALL = Instruction::DELEGATECALL;
-	static auto constexpr CREATE2 = Instruction::CREATE2;
-	static auto constexpr REVERT = Instruction::REVERT;
-	static auto constexpr INVALID = Instruction::INVALID;
-	static auto constexpr SELFDESTRUCT = Instruction::SELFDESTRUCT;
+
+	template<Instruction inst>
+	struct PatternGenerator
+	{
+		template<typename... Args> constexpr Pattern operator()(Args&&... _args) const
+		{
+			return {inst, make_vector<Pattern>(Pattern{std::forward<Args>(_args)}...)};
+		};
+	};
+
+	struct PatternGeneratorInstance
+	{
+		Instruction instruction;
+		PatternGeneratorInstance(Instruction _instruction): instruction(_instruction) {}
+		template<typename... Args> constexpr Pattern operator()(Args&&... _args) const
+		{
+			return {instruction, make_vector<Pattern>(Pattern{std::forward<Args>(_args)}...)};
```

Comment:
What failed with just ``{instruction, {std::forward<Args>(_args)}}`` or ``{instruction, std::forward<Args>(_args)}`` here? Every change I see here makes the compile-time-evaluation properties of this worse :-)...

Reply:
The original solution would sometimes invoke the `vector(size_t, T const&)`-constructor.

Reply:
Right - I should have thought about that one, that's a mean trap... What about changing the ``Pattern`` be constructed by an ``std::initializer_list<Pattern>`` instead of a ``std::vector<Pattern>``? Especially since the vector seems to be copied and not moved anyways...

Reply:
So my guess would be that the constructor can just as well construct the vector from an initializer list instead of copying a vector, which would avoid this ambiguity for all times... I haven't checked if a ``Pattern`` is ever actually constructed directly from a ``std::vector`` though... (which we could still have as another, but ``explicit`` constructor, if needed)

Reply:
Still does not compile on windows :(

Reply:
I think changing the arguments to the Pattern constructors to ``std::initializer_list<Pattern>``s would fix it...

Reply:
Together with removing all the ``make_vector`` stuff and moving back to plain brackets...